x/campaign/keeper/msg_server_add_mainnet_vesting_account.go

```diff
@@ -0,0 +1,75 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	spnerrors "github.com/tendermint/spn/pkg/errors"
+	"github.com/tendermint/spn/x/campaign/types"
+	profiletypes "github.com/tendermint/spn/x/profile/types"
+)
+
+func (k msgServer) AddMainnetVestingAccount(goCtx context.Context, msg *types.MsgAddMainnetVestingAccount) (*types.MsgAddMainnetVestingAccountResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	campaign, found := k.GetCampaign(ctx, msg.CampaignID)
+	if !found {
+		return nil, sdkerrors.Wrapf(types.ErrCampaignNotFound, "%v", msg.CampaignID)
+	}
+
+	// Get the coordinator ID
+	coordinatorID, found := k.profileKeeper.CoordinatorIDFromAddress(ctx, msg.Coordinator)
+	if !found {
+		return nil, sdkerrors.Wrap(profiletypes.ErrCoordAddressNotFound, msg.Coordinator)
+	}
+	if campaign.CoordinatorID != coordinatorID {
+		return nil, sdkerrors.Wrap(profiletypes.ErrCoordInvalid, fmt.Sprintf(
+			"coordinator of the campaign is %v",
+			campaign.CoordinatorID,
+		))
+	}
+
+	if campaign.MainnetInitialized {
+		return nil, sdkerrors.Wrapf(types.ErrMainnetInitialized, "%v", msg.CampaignID)
+	}
+
+	// check if the account already exists
+	oldAccount, foundAcc := k.GetMainnetVestingAccount(ctx, campaign.Id, msg.Address)
+	if foundAcc {
+		// if yes, remove the allocated share
+		totalShares, err := oldAccount.GetTotalShares()
+		if err != nil {
+			return nil, spnerrors.Critical(err.Error())
+		}
+		campaign.AllocatedShares, err = types.DecreaseShares(campaign.AllocatedShares, totalShares)
+		if err != nil {
+			return nil, sdkerrors.Wrapf(types.ErrTotalShareLimit, "%v", err.Error())
```

Comment:
The error here must be a crititcal error because we had more shares in accounts than the value in allocated shares